[AIE2P] Support wide vector postinc 2D/3D, and offset load/store

abnikant opened the pull request:
1. Enable combine_load_store_increment, combine_offset_load_store_ptradd and combine_offset_load_store_share_ptradd
2. Support wide vector POSTINC, POSTINC_2D, POSTINC_3D load and store.
3. Fixup , offset opcode.
4. Enable fifo combined load/store incr and offset load/store for 512-bits.
5. Add tests for combining and load/store.

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/combine-loads-stores.mir:
```diff
@@ -0,0 +1,1661 @@
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
nit: 2025

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -58,7 +58,10 @@ MachineInstr *findPreIncMatch(MachineInstr &MemI, MachineRegisterInfo &MRI,
   // No need to do it here.
-  if (MRI.getType(MemI.getOperand(0).getReg()).getSizeInBits() >= 1024)
+  MachineFunction &MF = *MemI.getMF();
+  bool isAIE2 = MF.getTarget().getTargetTriple().isAIE2();
```
nit: `const`

abnikant replied:
done

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -320,9 +323,11 @@ MachineInstr *findPostIncMatch(MachineInstr &MemI, MachineRegisterInfo &MRI,
     return nullptr;
-  if (MRI.getType(MemI.getOperand(0).getReg()).getSizeInBits() >= 1024)
+  MachineFunction &MF = *MemI.getMF();
+  bool isAIE2 = MF.getTarget().getTargetTriple().isAIE2();
```
nit: `const`

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/combine-loads-stores.mir:
```diff
@@ -0,0 +1,1661 @@
+  ; CHECK-NEXT: bb.1:
+  ; CHECK-NEXT:   successors: %bb.1(0x80000000)
+  ; CHECK-NEXT: {{  $}}
+  ; CHECK-NEXT:   [[PHI:%[0-9]+]]:_(p0) = G_PHI [[COPY]](p0), %bb.0, %2(p0), %bb.1
```
Interesting, here we have just `%2` instead of the regex.

khallouh commented on llvm/lib/Target/AIE/aie2p/AIE2PInstructionSelector.cpp:
```diff
@@ -1890,7 +1888,35 @@ LoadStoreOpcodes AIE2PInstructionSelector::getLoadStoreOpcode(
+        return {ISelOpcode, FitsImmediateRange,
+                /*OffsetOpcode=*/AIE2P::VLDA_dmx_lda_fifohl_idx_imm};
+      }
       llvm_unreachable("512-bit vector type must be in AccRegBank or VRegBank");
```
Nit: `...or FifoRegBank`

abnikant replied:
done.

khallouh commented on llvm/lib/Target/AIE/aie2p/AIE2PInstrInfo.cpp:
```diff
@@ -252,6 +252,11 @@ unsigned AIE2PInstrInfo::getOffsetMemOpcode(unsigned BaseMemOpcode) const {
   llvm_unreachable("not a generic load/store");
 }
 
+bool AIE2PInstrInfo::isOffsetMemOpcode(unsigned Opcode) const {
```
Nit: `isGenericMemOffsetOpcode`

abnikant replied:
updated the name.

martien-de-jong commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -58,7 +58,10 @@ MachineInstr *findPreIncMatch(MachineInstr &MemI, MachineRegisterInfo &MRI,
+  MachineFunction &MF = *MemI.getMF();
+  const bool isAIE2 = MF.getTarget().getTargetTriple().isAIE2();
+  if (isAIE2 &&
+      MRI.getType(MemI.getOperand(0).getReg()).getSizeInBits() >= 1024)
```
Shouldn't there be a corresponding size check for AIE2P?

abnikant replied:
yes, I have added a check but I don't expect the vector size to be greater that 2048-bits.

martien-de-jong commented on llvm/lib/Target/AIE/aie2p/AIE2PInstrInfo.cpp:
```diff
@@ -253,6 +253,11 @@ unsigned AIE2PInstrInfo::getOffsetMemOpcode(unsigned BaseMemOpcode) const {
 
+bool AIE2PInstrInfo::isGenericOffsetMemOpcode(unsigned Opcode) const {
+  return ((Opcode == AIE2P::G_AIE_OFFSET_STORE) ||
+          (Opcode == AIE2P::G_AIE_OFFSET_LOAD));
```
What about G_AIE_OFFSET_SEXTLOAD / G_AIE_OFFSET_ZEXTLOAD ?

abnikant replied:
done

F-Stuckmann commented on llvm/test/CodeGen/AIE/aie2p/fifo-loads.ll:
```diff
@@ -28,10 +28,13 @@ define dso_local void @_Z17test_fifo_ld_fillRPDv64_DB8_R12fifo_state_t(ptr nocap
-; CHECK-NEXT:    vst lfl0, [p1, #0] // Delay Slot 4
-; CHECK-NEXT:    vst lfh0, [p1, #64] // Delay Slot 3
+; CHECK-NEXT:    nop
+; CHECK-NEXT:    nop
```
why can the vst not work on the fifo registers as before?

abnikant replied:
not sure I understood the question but if your question is why the test is updated ? since MIR is updated after the load/store combiner is enabled, it uses G_AIE_POSTINC_STORE.

F-Stuckmann replied:
do we have a vst post increment that works on fifo regs? Because that would be preferable in this situation

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/ldst-fifo-stores.ll:
```diff
@@ -478,10 +490,12 @@ define dso_local void @test_fifo_st_push_v64bfp16ebs16(ptr nocapture nonnull ali
-; CHECK-NEXT:    vst sfl, [p1, #0] // Delay Slot 4
-; CHECK-NEXT:    vst sfh, [p1, #64] // Delay Slot 3
+; CHECK-NEXT:    nop
+; CHECK-NEXT:    vmov x1, sfh
```
I guess @F-Stuckmann is referring to this case. Now we are storing using vector registers.

abnikant replied:
we do have the post-increment for fifo-regs , refer to the test case inst-select-vector-pre-post-increment.mir. The problem with the above case is register bank assignment not assigning fiforegbank.


  %7:ptrregbank(p0), %8:fiforegbank(<32 x s32>), %9:gprregbank(s32) = G_INTRINSIC_W_SIDE_EFFECTS intrinsic(@llvm.aie2p.fifo.ld.fill), %4(p0), %5(<32 x s32>), %6(s32)
    %10:vregbank(<32 x s32>) = COPY %8(<32 x s32>)
    %3:ptrregbank(p0) = G_AIE_POSTINC_STORE %10(<32 x s32>), %1, %2(s20) :: (store (<32 x s32>) into %ir.s)

if I use %8 in G_AIE_POSTINC_STORE , then it generates post-inc with fifo. We are missing G_AIE_POSTINC_STORE opcode in regbank-select.

abnikant replied:
After handling G_AIE_POSTINC_STORE in RegisterBank assignment, both of these tests ldst-fifo-stores.ll and fifo-loads.ll are working as expected. I will have a separate commit for this.

andcarminati replied:
Hi @abnikant, I suspected that we were missing this proper register bank assignment. Thank you for the explanation.

abnikant replied:
@andcarminati I will open a separate PR for RegisterBank assignment, adding a commit in this PR does not make much sense.

andcarminati replied:
Agree! This PR has complexity enough.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -58,7 +58,11 @@ MachineInstr *findPreIncMatch(MachineInstr &MemI, MachineRegisterInfo &MRI,
+  const Triple &TT = MF.getTarget().getTargetTriple();
+  const unsigned VecSize =
+      MRI.getType(MemI.getOperand(0).getReg()).getSizeInBits();
+  if ((TT.isAIE2() && VecSize >= 1024) || (TT.isAIE2P() && VecSize > 2048))
```
nit: can we refactor this check? Future targets can just extend it.

abnikant replied:
I can refactor, and probably add a target hook to check for maximum supported size for a sub-target. But here we still need to add check for VecSize == 1024 for AIE2 since we don't have instruction selection support for 1024-bit combined load/store (postinc/offset)?

andcarminati replied:
Sure, any refactoring is welcome here ;-)

abnikant replied:
@andcarminati - Could you please review the refactored code? I had to merge two commits to prevent test failures during instruction selection after enabling combines, as support for instruction selection was introduced in the subsequent commit.

andcarminati replied:
Hi @abnikant, I had a second round or review, I included some suggestions to try to simplify some parts. The implementation looks correct to me.

abnikant committed:
1) [AIE2P] Enable post-pre incr and offset load combine
2) [AIE2P] Support postinc 2D/3D, and offset load/store

andcarminati commented on llvm/lib/Target/AIE/AIEBaseInstrInfo.h:
```diff
@@ -567,6 +569,11 @@ struct AIEBaseInstrInfo : public TargetInstrInfo {
 
+  /// Return the maximum supported vector size for this target.
+  virtual unsigned getMaxVectorBitSize() const {
+    llvm_unreachable("Target didn't implement getMaxVectorSize!");
```
nit `getMaxVectorBitSize`.

andcarminati commented on llvm/lib/Target/AIE/AIECombinerHelper.cpp:
```diff
@@ -58,8 +58,20 @@ MachineInstr *findPreIncMatch(MachineInstr &MemI, MachineRegisterInfo &MRI,
+  const unsigned VecSize =
+      MRI.getType(MemI.getOperand(0).getReg()).getSizeInBits();
+  const unsigned MaxVecSize = TII.getMaxVectorBitSize();
+  // TODO: Remove the following check once 1024-bit load/store
```
nit: outdated comment.

andcarminati replied:
I feel that we could have a common logic here, based on TII. Can we harmonize this asymmetric difference in some way? What do you think?

abnikant replied:
Can't skip this check if we use getMaxVectorBitSize , I have added another target hook getMaxSupportedLdStIncSize(), is this okay?

andcarminati commented on llvm/lib/Target/AIE/aie2p/AIE2PInstrInfo.cpp:
```diff
@@ -259,6 +259,13 @@ unsigned AIE2PInstrInfo::getOffsetMemOpcode(unsigned BaseMemOpcode) const {
   llvm_unreachable("not a generic load/store");
 }
 
+bool AIE2PInstrInfo::isGenericOffsetMemOpcode(unsigned Opcode) const {
```
Maybe `isAIEOffsetMemOpcode`?

abnikant replied:
I am not in favor of selectively updating the name, Generic keyword is used in few other places for AIE opcodes, I think updating all other names along with this will make more sense , what do you think (using a separate small PR) ?

andcarminati commented on llvm/lib/Target/AIE/aie2p/AIE2PInstructionSelector.cpp:
```diff
@@ -1926,7 +1952,8 @@ LoadStoreOpcodes AIE2PInstructionSelector::getLoadStoreOpcode(
       return {ISelOpcode, FitsImmediateRange,
               /*OffsetOpcode=*/AIE2P::VLDA_dmw_lda_w_idx_imm};
-    } else if (LoadStoreSize == 512) {
+    } else if (LoadStoreSize == 512 || LoadStoreSize == 1024 ||
```
This is a common predicate in several `ifs`. Do you think we can have a common predicate, like `isValidVectorRegSize`?

abnikant replied:
yes, this makes sense, but these are the ValidVectorRegSize (256, 512, 1024 and 2048-bits). There are special case handling where we don't have accumulator vector e.g. 256-bit , and also there are conditional block where we handle 512-bits. So, if we add  isValidVectorRegSize for few cases and somehow code is reordered then we will have problem ?

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/inst-select-vector-indexed-load-store.mir:
```diff
@@ -498,3 +498,936 @@ body:             |
+    $lfl0 = COPY %0:fiforegbank(<16 x s32>)
+...
+
+---
```
super nit: `underMinOffset`. I look to `overMin` and in my head I expect it to fit in the immediate ;-)

abnikant replied:
updated the name to outOfMinRangeOffset (Updated this file to include changes for other tests as well.)

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/ldst-fifo-stores.ll:
```diff
@@ -56,10 +56,12 @@ define dso_local void @_Z17test_fifo_st_pushRPDv64_DB8_S0_R12fifo_state_t(ptr no
-; CHECK-NEXT:    st r26, [p1, dj0] // Delay Slot 5
-; CHECK-NEXT:    vst sfl, [p1, #0] // Delay Slot 4
-; CHECK-NEXT:    vst sfh, [p1, #64] // Delay Slot 3
+; CHECK-NEXT:    nop
```
Maybe we could prioritize merging your RegBank PR first, as it represents an early compilation step. In this cases we will have this tests in the final shape. But it is just one idea.

abnikant replied:
That makes sense, but we need to merge this PR first; otherwise, `ld-fifo.ll` will break. After the Regbank PR, the regbank assignment for one of the operands in the load/store combine instruction is updated to `fiforegbank`, which isn't supported for instruction selection until this PR is merged.

andcarminati replied:
In this case, it makes sense!

andcarminati reviewed (dismissed):
LGTM. Nice work! Thank you for addressing the comments.

abnikant commented:
> LGTM. Nice work! Thank you for addressing the comments.

Thanks @andcarminati for looking into this. I just rebased this PR to resolve conflicts in fifo-loads.ll test. Can you please check ?

andcarminati commented on llvm/lib/Target/AIE/aie2p/AIE2PInstructionSelector.cpp:
```diff
@@ -1992,14 +2027,19 @@ LoadStoreOpcodes AIE2PInstructionSelector::getLoadStoreOpcode(
         return {/*ISelOpcode=*/AIE2P::VLDA_2D_dmx_lda_x, NoImmediate,
-                /*OffsetOpcode=*/{}};
-      llvm_unreachable("512-bit vector type must be in AccRegBank or VRegBank");
+                /*OffsetOpcode=*/{AIE2P::VLDA_dmw_lda_w_idx_imm}};
```
Late observation: is this `VLDA_dmw_lda_w_idx_imm` correct for size >= 512?

abnikant replied:
oh!, yes. This is wrong opcode and will result in wrong Copy. It should be VLDA_dmx_lda_x_idx_imm, I see it is used at two places. I will open a fixup PR shortly. Thanks for catching this.

andcarminati commented on llvm/test/CodeGen/AIE/aie2p/GlobalIsel/inst-select-vector-pre-post-increment.mir:
```diff
@@ -614,5 +614,1256 @@ body: |
+    ; CHECK-NEXT: [[MOV_PD_imm11_pseudo3:%[0-9]+]]:edc = MOV_PD_imm11_pseudo 4
+    ; CHECK-NEXT: [[REG_SEQUENCE:%[0-9]+]]:ed = REG_SEQUENCE [[MOV_PD_imm11_pseudo]], %subreg.sub_mod, [[MOV_PD_imm11_pseudo2]], %subreg.sub_dim_size, [[MOV_PD_imm11_pseudo1]], %subreg.sub_dim_stride, [[MOV_PD_imm11_pseudo3]], %subreg.sub_dim_count
+    ; CHECK-NEXT: [[VLDA_2D_dmx_lda_x:%[0-9]+]]:vec512, [[VLDA_2D_dmx_lda_x1:%[0-9]+]]:ep, [[VLDA_2D_dmx_lda_x2:%[0-9]+]]:edc = VLDA_2D_dmx_lda_x [[COPY]], [[REG_SEQUENCE]] :: (load (<16 x s32>), align 128)
+    ; CHECK-NEXT: [[VLDA_dmw_lda_w_idx_imm:%[0-9]+]]:mwa = VLDA_dmw_lda_w_idx_imm [[COPY]], 64 :: (load (<16 x s32>) from unknown-address + 64)
```
Here the same observation.
